File: rust/crates/phenix-conductor/src/bin/fixture-agent.rs
use agent_client_protocol::schema::v1::{
    AgentCapabilities, ClientRequest, ContentBlock, ContentChunk, InitializeResponse,
    PromptResponse, SessionNotification, SessionUpdate, StopReason, TextContent,
};
use agent_client_protocol::{Agent, Stdio};
use serde_json::{json, Value};
use std::error::Error;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<(), Box<dyn Error>> {
    Agent
        .builder()
        .name("phenix-conductor-fixture-agent")
        .on_receive_request(
            async move |request: ClientRequest, responder, connection| {
                let response = match request {
                    ClientRequest::InitializeRequest(initialize) => serde_json::to_value(
                        InitializeResponse::new(initialize.protocol_version)
                            .agent_capabilities(AgentCapabilities::new()),
                    )
                    .map_err(agent_client_protocol::Error::into_internal_error)?,
                    ClientRequest::NewSessionRequest(_) => json!({
                        "sessionId": "fixture-session",
                        "configOptions": fixture_config_options(),
                    }),
                    ClientRequest::SetSessionConfigOptionRequest(request) => {
                        let value = serde_json::to_value(&request.value)
                            .map_err(agent_client_protocol::Error::into_internal_error)?;
                        json!({
                            "configOptions": fixture_config_options_with(
                                &request.config_id.to_string(),
                                &value,
                            ),
                        })
                    }
                    ClientRequest::PromptRequest(prompt) => {
                        let text = prompt
                            .prompt
                            .into_iter()
                            .filter_map(|block| match block {
                                ContentBlock::Text(text) => Some(text.text),
                                _ => None,
                            })
                            .collect::<Vec<_>>()
                            .join("\n");
                        connection.send_notification(SessionNotification::new(
                            prompt.session_id,
                            SessionUpdate::AgentMessageChunk(ContentChunk::new(
                                ContentBlock::Text(TextContent::new(format!("echo: {text}"))),
                            )),
                        ))?;
                        serde_json::to_value(PromptResponse::new(StopReason::EndTurn))
                            .map_err(agent_client_protocol::Error::into_internal_error)?
                    }
                    _ => return Err(agent_client_protocol::Error::method_not_found()),
                };
                responder.respond(response)
            },
            agent_client_protocol::on_receive_request!(),
        )
        .connect_to(Stdio::new())
        .await?;
    Ok(())
}

fn fixture_config_options() -> Value {
    fixture_config_options_with("", &Value::Null)
}

fn fixture_config_options_with(config_id: &str, value: &Value) -> Value {
    let selected_model = if config_id == "model" {
        value.as_str().unwrap_or("provider/model")
    } else {
        "provider/model"
    };
    let selected_thinking = if config_id == "thinking" {
        value.as_str().unwrap_or("medium")
    } else {
        "medium"
    };
    json!([
        {
            "id": "model",
            "name": "Model",
            "category": "model",
            "type": "select",
            "currentValue": selected_model,
            "options": [{
                "value": "provider/model",
                "name": "Fixture Model",
            }],
        },
        {
            "id": "thinking",
            "name": "Thinking",
            "category": "thought_level",
            "type": "select",
            "currentValue": selected_thinking,
            "options": [
                { "value": "minimal", "name": "Minimal" },
                { "value": "low", "name": "Low" },
                { "value": "medium", "name": "Medium" },
                { "value": "high", "name": "High" },
                { "value": "max", "name": "Max" }
            ],
        }
    ])
}

File: nvim/tests/fixture_agent.py
#!/usr/bin/env python3
import json
import sys
import time


def send(message):
    sys.stdout.write(json.dumps(message, separators=(",", ":")) + "\n")
    sys.stdout.flush()


def response(request_id, result):
    send({"jsonrpc": "2.0", "id": request_id, "result": result})


for raw_line in sys.stdin:
    raw_line = raw_line.strip()
    if not raw_line:
        continue
    message = json.loads(raw_line)
    method = message.get("method")
    request_id = message.get("id")
    params = message.get("params") or {}

    if method == "initialize":
        response(
            request_id,
            {
                "protocolVersion": 1,
                "agentCapabilities": {},
            },
        )
    elif method == "_phenix/config/apply":
        response(
            request_id,
            {
                "revision": 1,
                "definition_id": params["input"]["definition_id"],
                "router": params["input"]["router"],
            },
        )
    elif method == "session/new":
        response(
            request_id,
            {
                "sessionId": "fixture-session",
                "configOptions": [
                    {
                        "id": "model",
                        "name": "Model",
                        "category": "model",
                        "type": "select",
                        "currentValue": "fixture-model",
                        "options": [
                            {"value": "fixture-model", "name": "Fixture Model"},
                            {"value": "other-model", "name": "Other Model"},
                        ],
                    }
                ],
            },
        )
    elif method == "session/set_config_option":
        response(
            request_id,
            {
                "configOptions": [
                    {
                        "id": params["configId"],
                        "name": "Model",
                        "category": "model",
                        "type": "select",
                        "currentValue": params["value"],
                        "options": [
                            {"value": "fixture-model", "name": "Fixture Model"},
                            {"value": "other-model", "name": "Other Model"},
                        ],
                    }
                ]
            },
        )
    elif method == "session/prompt":
        text = "\n\n".join(
            block.get("text", "")
            for block in params.get("prompt", [])
            if block.get("type") == "text"
        )
        send(
            {
                "jsonrpc": "2.0",
                "method": "session/update",
                "params": {
                    "sessionId": params["sessionId"],
                    "update": {
                        "sessionUpdate": "agent_thought_chunk",
                        "content": {"type": "text", "text": "thinking about: " + text},
                    },
                },
            }
        )
        if text == "scroll while streaming":
            time.sleep(0.25)
        send(
            {
                "jsonrpc": "2.0",
                "method": "session/update",
                "params": {
                    "sessionId": params["sessionId"],
                    "update": {
                        "sessionUpdate": "agent_message_chunk",
                        "content": {"type": "text", "text": "echo: " + text},
                    },
                },
            }
        )
        response(request_id, {"stopReason": "end_turn"})
    elif method == "session/close":
        response(request_id, {})
    elif request_id is not None:
        send(
            {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {"code": -32601, "message": "unknown method: " + str(method)},
            }
        )
